### ISSUE TITLE
Allow using RS0034 without flagging default constructors

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/roslyn-analyzers#3025]
#### Analyzer package

[Roslyn.Diagnostics.Analyzers](https://www.nuget.org/packages/Roslyn.Diagnostics.Analyzers)

#### Package Version

v2.9.7

#### Diagnostic ID

RS0034 Exported parts should have [ImportingConstructor]

#### Repro steps

In the project system repo, we don't have RS0034 turned on. I wasted some time because I missed an `[ImportingConstructor]` on a non-default constructor. I wanted to turn this analyzer on. Doing so triggers 51 warnings about default constructors, none of which are real warnings.

This is not a problem, but triggers RS0034:

```c#
[Export(typeof(IFoo))]
public class Foo : IFoo
{
   void IFoo.Bar() {}
}
```

This is a problem, and triggers RS0034:

```c#
[Export(typeof(IFoo))]
public class Foo : IFoo
{
   public Foo(IBaz baz) {}
   void IFoo.Bar() {}
}
```

#### Expected behavior

Have the ability to to have diagnostics on the second case _only_.

#### Actual behavior

Diagnostics on both cases, however the first case is not an actual product issue.

---

I understand some people will like to be explicit that a type has an importing constructor. My perspective is that when a real problem is introduced then the RS0034 diagnostic would still trigger (indicating a real problem). That for me is the sweet spot. Can this either be configured or split into two diagnostics?

**Comment by Evangelink** [comment]
@mavasani Any preference between:
1. New analyzer
2. Editor option
3. Simply changing the rule

**Comment by mavasani** [comment]
Tagging @sharwell - I'd go with his recommendation here.

**Comment by sharwell** [comment]
The first case is a problem because the implicit constructor is missing the `Obsolete` attribute that each MEF part should have to avoid direct construction in test code. This analyzer is really geared toward dotnet/roslyn; I'll need to think about what might be done to support other repositories.

**Comment by Evangelink** [comment]
@sharwell ping

**Comment by sharwell** [comment]
Like all rules in Roslyn.Diagnostics.Analyzers, this analyzer is not intended for use outside of dotnet/roslyn. I would be in favor of creating a new set of MEF rules for a broader audience as part of the more widely used API analysis package(s), but I don't see a need to change a dotnet/roslyn rule so it works in a case it was never intended.
